Transaction Part 4: LockManager, CycleDetector and DeadLockDetector.

hakanmemisoglu opened the pull request:
This PR introduces the last piece of the transaction module:

- LockManager: This is the entry point for the module. All outside interaction will happen via this class.

- DeadLockDetector: This class contains the logic for deadlock detection and selection of victims.

- CycleDetector: This class will be used by DeadLockDetector for detection of the cycles in the wait-for graph.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace transaction {
+
+CycleDetector::CycleDetector(DirectedGraph *wait_for_graph)
+  : wait_for_graph_(wait_for_graph) {
```
According to the code style in `Line length` part, please add two more whitespace indentations.

https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+
+CycleDetector::CycleDetector(DirectedGraph *wait_for_graph)
+  : wait_for_graph_(wait_for_graph) {
+  scc_ = std::make_unique<StronglyConnectedComponents>(*wait_for_graph_);
```
We could move this assignment in the constructor's init list.

```
  wait_for_graph_(wait_for_graph),
  scc_(new StronglyConnectedComponents(*wait_for_graph_)) {
```

Btw, could we have a better name for `scc`? The full name will be better.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+  std::unordered_map<std::uint64_t, std::vector<DirectedGraph::node_id>>
+    component_mapping = scc_->getComponentMapping();
+  for (std::pair<std::uint64_t, std::vector<DirectedGraph::node_id>>
+         entry : component_mapping) {
```
Same style issues: total 4 space indentation.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+      for (DirectedGraph::node_id elem : targets) {
+        nodes_to_kill.push_back(elem);
+      }
+    }
```
We could refactor to remove one level indentation:

```
  if (...) {
    continue;
  }

  const std::vector<DirectedGraph::node_id> nodes
      = breakComponent(entry.second);
  for (DirectedGraph::node_id node : nodes) {
      nodes_to_kill.push_back(node);
  }
```

Note that I also rename `targets` to `nodes`, as well as `elem` to `node`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+CycleDetector::breakComponent(const std::vector<DirectedGraph::node_id> &nodes) {
+  std::vector<DirectedGraph::node_id> targets;
+  // Convert it to set to ensure defensively that the elements are unique.
+  std::set<DirectedGraph::node_id> nodes_set(nodes.begin(), nodes.end());
```
Could we use `unordered_set` instead?

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+  // Convert it to set to ensure defensively that the elements are unique.
+  std::set<DirectedGraph::node_id> nodes_set(nodes.begin(), nodes.end());
+  while (true) {
+    bool has_cycle = hasCycleWithin(nodes_set);
```
Please mark `const`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+    // TODO(Hakan): This is very inefficient scheme, however in the
+    //              future, we can use the transaction's priority
+    //              as the victim selection parameter.
+    DirectedGraph::node_id victim = *(nodes_set.begin());
```
Please mark `const`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+    to_visit.push(node_id);
+    // Start to visit nodes until it is done.
+    while (!to_visit.empty()) {
+      DirectedGraph::node_id current_node = to_visit.top();
```
Please mark `const`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+      // Mark the node coming from stack as "visited".
+      visited.insert(current_node);
+      // For all adjacent nodes of this "visited" node,
+      std::vector<DirectedGraph::node_id> adjacents
```
Please mark `const`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+CycleDetector::hasCycleWithin(const std::set<DirectedGraph::node_id> &within) {
+  // Keeps track of the nodes the algorithms visited.
+  std::set<DirectedGraph::node_id> visited;
+  for (DirectedGraph::node_id node_id : within) {
```
Add `const`: `for (const DirectedGraph::node_id node_id : within) {`.

zuyu commented:
@hakanmemisoglu I have not done one pass, but please change `set` to `unordered_set` for performance reasons.

hakanmemisoglu commented:
Thanks @zuyu,
I fixed as you suggested.

pateljm commented:
@zuyu -- shall we merge this? Thanks!

zuyu commented:
I'll take a look later today.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+  const std::unordered_map<std::uint64_t, std::vector<DirectedGraph::node_id>>
+    component_mapping = strongly_connected_components_->getComponentMapping();
+  for (const std::pair<std::uint64_t, std::vector<DirectedGraph::node_id>>
+           &entry : component_mapping) {
```
We'd better use `for (const auto &entry : component_mapping)`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+        = breakComponent(entry.second);
+      for (const DirectedGraph::node_id &node : nodes) {
+        nodes_to_kill.push_back(node);
+      }
```
I was wondering why the `for` loop has an extra level indentation.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+
+  while (true) {
+    const bool has_cycle = hasCycleWithin(nodes_set);
+    if (!has_cycle) {
```
Since `has_cycle` does not use in anywhere else, how about the following refactor:

```
  if (!hasCycleWithin(nodes_set)) {
    break;
  }
```

Also, regarding the method name, I prefer `hasCycle` instead of `hasCycleWithin`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+    // TODO(Hakan): This is very inefficient scheme, however in the
+    //              future, we can use the transaction's priority
+    //              as the victim selection parameter.
+    const DirectedGraph::node_id victim = *(nodes_set.begin());
```
Add a new method called `chooseVictim` with the naive implementation you have.

zuyu commented on transaction/CycleDetector.hpp:
```diff
@@ -0,0 +1,75 @@
+      breakComponent(const std::vector<DirectedGraph::node_id> &nodes);
+
+  // Checks whether the nodes in the set make a cycle.
+  bool hasCycleWithin(const std::unordered_set<DirectedGraph::node_id> &within);
```
I was wondering why choose `within` as the argument name.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+      // For all adjacent nodes of this "visited" node,
+      const std::vector<DirectedGraph::node_id> adjacents
+          = wait_for_graph_->getAdjacentNodes(current_node);
+      for (DirectedGraph::node_id adj : adjacents) {
```
Please add `const` for `adj`. Also, rename it to `adjacent`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,122 @@
+          // If this adjacent node is a node we already visited, then
+          // there is a cycle.
+          return true;
+        } else if (within.count(adj) == 1 && visited.count(adj) == 0) {
```
I was wondering if `within.count(adj) == 1` is always `true`.

hakanmemisoglu replied:
It won't be the case. Sometimes it is `within.count(adj) == 0` , and it means iterator is right now, does not point the node which is in first set (SCC). Then there is no point to continue with this path. That's why sometimes this for loop won't do anything.

zuyu commented on transaction/CycleDetector.hpp:
```diff
@@ -0,0 +1,75 @@
+}  // namespace transaction
+}  // namespace quickstep
+
+#endif
```
Change to `#endif  // QUICKSTEP_TRANSACTION_CYCLE_DETECTOR_HPP_`.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,153 @@
+}  // namespace transaction
+}  // namespace quickstep
+
+#endif
```
Add `  // QUICKSTEP_TRANSACTION_DEAD_LOCK_DETECTOR_HPP_` in the end.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+namespace quickstep {
+namespace transaction {
+
+constexpr std::int64_t DeadLockDetector::kSleepDuration;
```
I was wondering why need this?

hakanmemisoglu replied:
Otherwise, GCC produces a linkage error.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+DeadLockDetector::DeadLockDetector(LockTable *lock_table,
+                                   std::atomic<DeadLockDetectorStatus> *status,
+                                   std::vector<DirectedGraph::node_id> *victims)
+    : wait_for_graph_(nullptr),
```
No need for initializing `wait_for_graph_`.

hakanmemisoglu replied:
I assume that making it explicit does not hurt.

zuyu replied:
But it's unnecessary.

hakanmemisoglu replied:
Ok, I will change it.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,153 @@
+   *        necessary information.
+   * @param status Pointer to status object which will act as a message
+   *        passing algorithm between LockManager.
+   * @paramsstd::atomic<DeadLockDetectorStatus> &status_;
```
?

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+DeadLockDetector::DeadLockDetector(LockTable *lock_table,
+                                   std::atomic<DeadLockDetectorStatus> *status,
+                                   std::vector<DirectedGraph::node_id> *victims)
```
I was wondering why passing these pointers?

hakanmemisoglu replied:
Do you want to convert them to reference?

zuyu replied:
Yes, if you don't modify them in this method.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,153 @@
+  std::vector<transaction_id> getAllVictims();
+
+ private:
+  static constexpr std::int64_t kSleepDuration = 5;
```
How about `kSleepDurationInSeconds`?

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+std::vector<transaction_id>
+DeadLockDetector::getAllDependents(transaction_id owner) {
+  FATAL_ERROR("Not implemented");
```
In all files, please use `LOG(FATAL)` in `glog`, instead `FATAL_ERROR`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+      // DeadLockDetector should stop.
+      return;
+    }
+    while (status_.load() == DeadLockDetectorStatus::kDone) {
```
Could you explain the logics here?

hakanmemisoglu replied:
DeadLockDetector will run once in 5 seconds. First it will check the buffer that it has filled is empty of not. That line is for checking this condition.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+void DeadLockDetector::addPendingInfo(transaction_id pending,
+                                      transaction_id owner) {
+  DirectedGraph::node_id pending_node_id = getNodeId(pending);
+  DirectedGraph::node_id owner_node_id = getNodeId(owner);
```
Please mark above two `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+}
+
+void DeadLockDetector::addPendingInfo(transaction_id pending,
+                                      transaction_id owner) {
```
Please mark two arguments `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  std::vector<transaction_id> transactions;
+  transactions.reserve(nodes.size());
+  for (DirectedGraph::node_id node_id : nodes) {
+    transaction_id tid = wait_for_graph_->getDataFromNode(node_id);
```
Please mark `const` in this method, when possible, from the argument to this line.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  std::vector<DirectedGraph::node_id> nodes
+      = wait_for_graph_->getAdjacentNodes(pending_node_id);
+  std::vector<transaction_id> transactions;
+  transactions.reserve(nodes.size());
```
I don't understand why you first reserve some places, but then append more elements in the end.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  return transactions;
+}
+
+DirectedGraph::node_id DeadLockDetector::getNodeId(transaction_id tid) {
```
Please mark the argument `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+
+DirectedGraph::node_id DeadLockDetector::addNode(transaction_id tid) {
+  transaction_id *tid_ptr = new transaction_id(tid);
```
I was wondering which is responsible to delete `transaction_id`?

hakanmemisoglu replied:
Do you mind to look at DirectedGraph?

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+DirectedGraph::node_id DeadLockDetector::addNode(transaction_id tid) {
+  transaction_id *tid_ptr = new transaction_id(tid);
+  DirectedGraph::node_id node_id = wait_for_graph_->addNodeUnchecked(tid_ptr);
```
Please mark `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+DirectedGraph::node_id DeadLockDetector::addNode(transaction_id tid) {
+  transaction_id *tid_ptr = new transaction_id(tid);
+  DirectedGraph::node_id node_id = wait_for_graph_->addNodeUnchecked(tid_ptr);
+  (*tid_node_mapping_)[tid] = node_id;
```
For performance reasons, please use `emplace` instead of the current implementation to avoid to set a default value and then another assignment.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+std::vector<transaction_id> DeadLockDetector::getAllVictims() {
+  std::vector<transaction_id> result_victims;
+  wait_for_graph_ = std::make_unique<DirectedGraph>();
```
I was wondering who is responsible to initialize `wait_for_graph_`. I know it is not in the constructor, but do we intend to delete the old graph, and create a new, empty one, as you has done here.

For implementation, I'd use `wait_for_graph_.reset(std::make_unique<DirectedGraph>());`, if you mean to initialize it here.

hakanmemisoglu replied:
Wait for graph is initialized and controlled by DeadLockDetector.
GetAllVictims called once in 5 seconds. It will be deleted and reconstructed.
Thank for suggestion. Reset call is more explicit.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+      transaction_id owned_transaction = it_own_list->first;
+
+      // TODO(Hakan): Convert them to log messages.
+      // std::cout << "Owned: " << owned_transaction << std::endl;
```
Please delete commented code.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  lock_table_->latchShared();
+
+  for (LockTable::const_iterator it = lock_table_->begin();
+       it != lock_table_->end(); ++it) {
```
I prefer to `for (const auto &lock_info : lock_table_)`.

hakanmemisoglu replied:
It is not standard library data structure. I need to write additional methods to support this syntactic sugar. Let's keep it as it is.

zuyu replied:
No, since we already have `begin` and `end`, the following code works on my Mac with `LLVM`:

```
  for (const auto &lock_item : *lock_table_) {
    const LockTable::lock_own_list &own_list = lock_item.second.first;
    const LockTable::lock_pending_list &pending_list = lock_item.second.second;

    ...
    }
  }
```

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+    const LockTable::lock_pending_list &pending_list = it->second.second;
+
+    for (LockTable::lock_own_list::const_iterator it_own_list = own_list.begin();
+         it_own_list != own_list.end(); ++it_own_list) {
```
I prefer `for (const auto &lock_pair : own_list)`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    for (LockTable::lock_own_list::const_iterator it_own_list = own_list.begin();
+         it_own_list != own_list.end(); ++it_own_list) {
+      transaction_id owned_transaction = it_own_list->first;
```
Please mark `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+      // TODO(Hakan): Convert them to log messages.
+      // std::cout << "Owned: " << owned_transaction << std::endl;
+
+      DirectedGraph::node_id owned_node = getNodeId(owned_transaction);
```
Please mark `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+      for (LockTable::lock_pending_list::const_iterator
+               it_pending_list = pending_list.begin();
+           it_pending_list != pending_list.end(); ++it_pending_list) {
```
I prefer a `for-each` loop.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+        transaction_id pending_transaction = it_pending_list->first;
+        // std::cout << "Pending: " <<  pending_transaction << std::endl;
+
+        DirectedGraph::node_id pending_node = getNodeId(pending_transaction);
```
Please mark above two `const`, and delete commented code.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+        wait_for_graph_->addEdgeUnchecked(pending_node, owned_node);
+
+        // std::cout << "AddEdge(" << pending_transaction << ", "
+        //           << owned_transaction << ")" << std::endl;
```
Please remove commented code.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  // Critical region on LockTable ends here.
+
+  CycleDetector cycle_detector(wait_for_graph_.get());
+  std::vector<DirectedGraph::node_id> victim_nodes = cycle_detector.breakCycle();
```
Mark `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  CycleDetector cycle_detector(wait_for_graph_.get());
+  std::vector<DirectedGraph::node_id> victim_nodes = cycle_detector.breakCycle();
+  for (DirectedGraph::node_id node_id : victim_nodes) {
```
Add `const` for `ndoe_id`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+  CycleDetector cycle_detector(wait_for_graph_.get());
+  std::vector<DirectedGraph::node_id> victim_nodes = cycle_detector.breakCycle();
+  for (DirectedGraph::node_id node_id : victim_nodes) {
+    transaction_id victim_tid = wait_for_graph_->getDataFromNode(node_id);
```
Mark `const`.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  // Destroy the wait graph. It will be reconstructed kSleepDurationSeconds
+  // seconds later.
+  wait_for_graph_.release();
```
This is a memory leak bug. You should call `reset` instead.

hakanmemisoglu replied:
Good catch, thanks.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,153 @@
+
+  DirectedGraph::node_id getNodeId(transaction_id tid);
+
+  DirectedGraph::node_id addNode(transaction_id tid);
```
In this file, please mark arguments `const` when possible.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,153 @@
+  LockTable *lock_table_;
+
+  std::atomic<DeadLockDetectorStatus> &status_;
+  std::vector<DirectedGraph::node_id> &victims_;
```
I was wondering if you ever changed `victims_` in this class.

hakanmemisoglu replied:
It is there, please look at the swap call at DeadLockDetector::run() method.

zuyu replied:
Then, we should explicitly use a pointer, instead of a reference.

hakanmemisoglu replied:
Ok, I will change it.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+          bool result = releaseAllLocks(request.getTransactionId());
+          if (!result) {
+            LOG(FATAL) << "Unexpected condition occured.";
+          }
```
Please refactor as following:

```
  CHECK(releaseAllLocks(request.getTransactionId()))
      << "Unexpected condition occured.";
```

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+            LOG(FATAL) << "Unexpected condition occured.";
+          }
+        } else {
+          bool result = acquireLock(request.getTransactionId(),
```
Please mark `const`. Alternatively, just refactor to
```
if (acquireLock(...)) {
  ...
} else {
  ...
}
```

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+  deadlock_detector_->start();
+
+  constexpr std::uint64_t max_try_incoming = 10000;
+  constexpr std::uint64_t max_try_inner = 6000;
```
Please rename to `kMaxTryIncoming` and `kMaxTryInner`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    }
+    for (std::uint64_t tries = 0; tries < max_try_inner; ++tries) {
+      if (!inner_pending_requests_.empty()) {
+        LockRequest request = inner_pending_requests_.front();
```
Could we mark `request` `const`?

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+      if (!inner_pending_requests_.empty()) {
+        LockRequest request = inner_pending_requests_.front();
+
+        bool result = acquireLock(request.getTransactionId(),
```
We could either mark `const`, or refactor to `if (acquireLock(...))`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+      (current_access_mode.isShareLock() ||
+       current_access_mode.isIntentionShareLock())
+      ? AccessMode(AccessModeType::kIsLock)
+      : AccessMode(AccessModeType::kIxLock);
```
According to the code style for breaking a line, we need two more whitespace indentations:

https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+  while (!stack.empty()) {
+    std::pair<ResourceId, AccessMode> pair_to_pick = stack.top();
+    ResourceId rid_to_pick = pair_to_pick.first;
+    AccessMode access_mode_to_pick = pair_to_pick.second;
```
Please mark all three above `const`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    ResourceId rid_to_pick = pair_to_pick.first;
+    AccessMode access_mode_to_pick = pair_to_pick.second;
+
+    bool result = acquireLockInternal(tid,
```
We could either mark `const`, or refactor `if (!acquireLockInternal(...))`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+      = transaction_table_->deleteTransaction(tid);
+  if (transaction_deleted
+          == TransactionTableResult::kTransactionDeleteError) {
+    FATAL_ERROR("In LockManager.releaseAllLocks"
```
Please replace `FATAL_ERROR` with `LOG(FATAL)` in `glog`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+  std::vector<ResourceId> related_rids
+      = transaction_table_->getResourceIdList(tid);
+  TransactionTableResult transaction_deleted
+      = transaction_table_->deleteTransaction(tid);
```
We could mark two above `const`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+}
+
+// If not blocked return true
+// if block return false
```
Remove these comments. They should appear in the method declaration in the header file.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+       ++it) {
+    LockTableResult lock_deleted = lock_table_->deleteLock(tid, *it);
+
+    std::cout << "Transaction " + std::to_string(tid)
```
No `cout` in all the codebase, except for any `main` methods.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+
+  for (std::vector<ResourceId>::const_iterator it = related_rids.begin();
+       it != related_rids.end();
+       ++it) {
```
How about a `for-each`: `for (const auto &resource_id : related_rids)`?

Also, rename `related_rids` to `resource_ids`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+bool LockManager::acquireLockInternal(const transaction_id tid,
+                                      const ResourceId &rid,
+                                      const AccessMode access_mode) {
+  LockTableResult l_result = lock_table_->putLock(tid, rid, access_mode);
```
How about `lock_result`?

When naming a variable, you should consider the case where you could recognize its meaning without thinking after half a year later reading the code again.

hakanmemisoglu replied:
Really?

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+  LockTableResult l_result = lock_table_->putLock(tid, rid, access_mode);
+  if (l_result == LockTableResult::kPutError) {
+    LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal";
+  }
```
Refactor to

```
  CHECK(l_result != LockTableResult::kPutError)
      << "Unexpected result in LockManager.acquireLockInternal";
```

If you wonder how `CHECK` works, read `glog` document. Basically, it always executes and expects `true`. I like `CHECK` rather than an `if` + `LOG(FATAL)`, because the former is simple and does not need further thinking its logic.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal";
+  }
+
+  if (l_result == LockTableResult::kAlreadyInOwned) {
```
How about refactor using a `switch` statement?

hakanmemisoglu replied:
LOG(FATAL) breaks the switch statement. Let's keep it as it is.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+  } else if (l_result == LockTableResult::kPlacedInOwned) {
+    TransactionTableResult t_result
+        = transaction_table_->putOwnEntry(tid, rid, access_mode);
+    if (t_result != TransactionTableResult::kPlacedInOwned) {
```
How about using `CHECK` here?

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+      transaction_table_->putPendingEntry(tid, rid, access_mode);
+    if (t_result != TransactionTableResult::kPlacedInPending) {
+      LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal: "
+                    "Mismatch of table results (No.2).";
```
Ditto for `CHECK`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    for (std::vector<transaction_id>::const_iterator iter
+             = victim_result_.begin();
+         iter != victim_result_.end();
+         ++iter) {
```
How about a `for-each`: `for (const auto victim_transaction_id : victim_result_)`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    : lock_table_(std::make_unique<LockTable>()),
+      transaction_table_(std::make_unique<TransactionTable>()),
+      detector_status_(DeadLockDetectorStatus::kNotReady),
+      victim_result_(),
```
Remove this line.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,235 @@
+
+void LockManager::killVictims() {
+  // TODO(Hakan): Find a method to latch this function
+  //              (it cannot because it calls releaseLocks)
```
You may want to add a mutex to protect `victim_result_`, and pass the mutex to `DeadlockDetector`.

hakanmemisoglu replied:
detector_status_ value will prevent the access of LockManager and DeadLockDetector's concurrent access to victim_result_ . Do you see any problem?

zuyu replied:
Well, I was thinking to fix the `TODO`. Any proposal here?

zuyu commented on transaction/LockManager.hpp:
```diff
@@ -0,0 +1,118 @@
+}  // namespace transaction
+}  // namespace quickstep
+
+#endif
```
Add `  // QUICKSTEP_TRANSACTION_LOCK_MANAGER_HPP_` in the end.

zuyu commented on transaction/LockTable.hpp:
```diff
@@ -40,14 +40,14 @@ namespace transaction {
+  kAlreadyInPending,
+  kDelFromOwned,
+  kDelFromPending,
+  kDelError,
```
How about changing `Del` to `Delete`.

zuyu commented on transaction/ResourceId.hpp:
```diff
@@ -202,10 +202,10 @@ class ResourceId {
+  database_id db_id_;
+  relation_id rel_id_;
+  block_id block_id_;
+  tuple_id tuple_id_;
```
I was wondering why dropped `const` here.

hakanmemisoglu replied:
= operator changes it.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,128 @@
+class CycleDetectorTest : public testing::Test {
+ protected:
+  virtual void SetUp() {
+    for (std::uint64_t i = 0; i < 12; ++i) {
```
I was wondering what the magic number `12` means.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,128 @@
+      transactions_.push_back(new transaction_id(i));
+    }
+
+    wait_for_graph_ = std::make_unique<DirectedGraph>();
```
Use `reset`.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,128 @@
+              {8, 6},
+              {9, 8}, {9, 10},
+              {10, 11},
+              {11, 9}};
```
Could you please add comments about the existing cycles, if any?

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,128 @@
+              {11, 9}};
+
+    for (std::uint64_t i = 0; i < edges_.size(); ++i) {
+      std::pair<DirectedGraph::node_id, DirectedGraph::node_id> edge = edges_[i];
```
Mark `const`.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,128 @@
+  std::unique_ptr<CycleDetector> cycle_detector_;
+};
+
+class CycleDetectorTestSimple : public testing::Test {
```
It seems that the only difference between two test classes are the edge info. Instead of adding `CycleDetectorTestSimple`, you could change `edges_` in `TEST_F(CycleDetectorTestSimple, BreakCycleSimple)`.

For more info, read the gtest documentation:

https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests

hakanmemisoglu replied:
More than half of the action happens after edges is initialized. If I do it same, I should call all these methods for myself. Let's keep it as it is.

zuyu replied:
The goal is to remove duplicate codes when possible. In this case, two test classes are almost the same, except for the edge set up, right?

hakanmemisoglu replied:
Duplicate codes are problem in the main code, I totally understand that. However, as you made the statement, these are tests. I really appreciate that I would have more freedom how to write unit tests. Moreover, the verbosity in each test makes the scenario more explicit. If the last lines of SetUp method were different, I could refactor that part into one class and add it as a statement of fixture tests. In that case, I could not. However, what you want is a configuration class that sets up the edges based on a vector pairs, then I can do that.

zuyu replied:
Well, there are multiple ways to write a `gtest`-based unit tests.

And I think you could write a better unit test using one single test class. If you need an example, check out `catalog/tests/Catalog_unittest.cpp` for initializing multiple cases.

FYI, we do not need to do everything in `SetUp`. So, a different unit test could have its own things for additional set up.

zuyu commented on transaction/tests/DeadLockDetector_unittest.cpp:
```diff
@@ -0,0 +1,76 @@
+
+class DeadLockDetectorTest : public ::testing::Test {
+ protected:
+  virtual void SetUp() {
```
You could use the constructor, instead of `SetUp`.

hakanmemisoglu replied:
I want SetUp.

zuyu replied:
I was wondering why you want to stick with `SetUp`:

```
  DeadLockDetectorTest()
      : lock_table_(std::make_unique<LockTable>()),
        status_(DeadLockDetectorStatus::kDone) {}
```

zuyu commented on transaction/tests/DeadLockDetector_unittest.cpp:
```diff
@@ -0,0 +1,76 @@
+
+  lock_table_->putLock(transaction_id(2),
+                       ResourceId(4, 5),
+                       AccessMode(AccessModeType::kXLock));
```
Please add comments explaining the resource graph you have.

hakanmemisoglu replied:
Not needed thanks.

zuyu replied:
Yes, we do need comments. Otherwise, how does the reader know what deadlock we are trying to expose, without digging into the code further?

hakanmemisoglu replied:
What I dismiss is unnecessary commenting, especially at unit test. It is obvious that this a schedule of write requests. What do you need more to understand the cycle structure here?

zuyu replied:
Well, I think we could refactor the code more clearly.

```
  const transaction_id transaction_one(1), transaction_two(2).
  const ResourceId resource_one(...), resource_two(...);
  const AccessMode x_lock(AccessModeType::kXLock);
  
  // Create an interleaving deadlock among two transactions and two resources
  // where each transaction holds one resource, and tries to acquire the other resource.
  lock_table_->putLock(transaction_one, resource_one, x_lock);
  lock_table_->putLock(transaction_two, resource_two, x_lock);
  
  lock_table_->putLock(transaction_one, resource_two, x_lock);
  lock_table_->putLock(transaction_two, resource_one, x_lock);
```

The code might not compile, but you should get the sense that we should explicitly express the interleaving deadlock by using variable names and how to add comments.

zuyu commented:
@hakanmemisoglu I've done one pass.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+      releaseAllLocks(victim_transaction_id);
+      // TODO(Hakan): Find a way to kill transaction, so that requests with this
+      //              tid should be ignored.
+      LOG(INFO) << "Killed transaction " + std::to_string(victim_transaction_id);
```
We may want to use `<<` instead of `+` before `std::string` to avoid temp string objects.

hakanmemisoglu replied:
My fear is that there would be interleaving of log messages by different thread in your suggestion. Does it has any protection for that particular problem?
For example:
"
Killed transaction SOME_OTHER_LOG //newline
TID:3456
"

zuyu replied:
Check out `glog` for the thread safe options.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,118 @@
+std::vector<DirectedGraph::node_id> CycleDetector::breakCycle() const {
+  std::vector<DirectedGraph::node_id> nodes_to_kill;
+  const std::unordered_map<std::uint64_t, std::vector<DirectedGraph::node_id>>
+    component_mapping = strongly_connected_components_->getComponentMapping();
```
According to the `line length` part in the code style, breaking a line needs four whitespace indentations. So in this case, we need two more.

https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide

zuyu commented on transaction/CycleDetector.hpp:
```diff
@@ -0,0 +1,80 @@
+   *
+   * @return Vector of node ids that should be killed to break all cycles.
+   */
+  std::vector<DirectedGraph::node_id> breakCycle() const;
```
We may want to choose a better name like `chooseVictimNode`. `breakCycle` means we do modify the graph to eliminate cycles.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,118 @@
+    //              future, we can use the transaction's priority
+    //              as the victim selection parameter.
+    const DirectedGraph::node_id victim = chooseVictim(nodes_set);
+    nodes_set.erase(nodes_set.begin());
```
We should change `nodes_set.begin()` to `victim`.

Also, move the above `TODO` inside `chooseVictim`.

zuyu commented on transaction/CycleDetector.cpp:
```diff
@@ -0,0 +1,118 @@
+    if (!hasCycle(nodes_set)) {
+      break;
+    }
+    // If there is a cycle, start to pop a node from randomly.
```
We may want to change the comment to `If there is a cycle, choose a victim to break a cycle.`

zuyu commented on transaction/CycleDetector.hpp:
```diff
@@ -0,0 +1,80 @@
+  breakComponent(const std::vector<DirectedGraph::node_id> &nodes) const;
+
+  inline DirectedGraph::node_id chooseVictim(
+      const std::unordered_set<DirectedGraph::node_id> nodes_set) const {
```
We should use a reference (`&`) here to avoid a deep copy.

zuyu commented on transaction/CycleDetector.hpp:
```diff
@@ -0,0 +1,80 @@
+
+ private:
+  std::vector<DirectedGraph::node_id>
+  breakComponent(const std::vector<DirectedGraph::node_id> &nodes) const;
```
We may also want to choose a better name like `chooseVictimsInComponent` for this method, as it does not break things.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,178 @@
+    status_->store(DeadLockDetectorStatus::kDone);
+
+    // DeadLockDetector should run once in a predefined interval.
+    std::this_thread::sleep_for(
```
Please include `<thread>` for what you use.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,178 @@
+
+
+DirectedGraph::node_id DeadLockDetector::addNode(const transaction_id tid) {
+  transaction_id *tid_ptr = new transaction_id(tid);
```
This question traces back to an old PR: I know `transaction_id` is an integer type, so I don't understand why we need to store it in a `unique_ptr` in `DirectedGraphNode`?

hakanmemisoglu replied:
I was trying to imitate the structure in DAG.hpp. However you are right in this case, it might better to store variable directly.

zuyu replied:
I see. The reason why `DAG.hpp` uses a `unique_ptr` is that we store objects. We should fix it in a separate PR.

And it is up to you regarding the order of merging two PRs. We have two options:
 1. We could merge another PR for storing `transaction_id` as an integer directly, and then rebasing this PR with minor changes.
 2. We could merge this PR first, and then do another PR to fix `transaction_id` usages in one shot.

zuyu commented on transaction/DeadLockDetector.cpp:
```diff
@@ -0,0 +1,178 @@
+  // Critical region on LockTable starts here.
+  lock_table_->latchShared();
+
+
```
Please remove this empty line.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+namespace quickstep {
+namespace transaction {
+
+LockManager::LockManager(ThreadSafeQueue<LockRequest> *incoming_requests,
```
Please include `utility/ThreadSafeQueue.hpp` for what you use.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+        } else {
+            LOG(INFO) << "Transaction "
+                         + std::to_string(request.getTransactionId())
+                         + " acquired " + request.getResourceId().toString();
```
Please use `<<` instead of `+` to avoid temp string objects.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+  deadlock_detector_->start();
+
+  constexpr std::uint64_t kMaxTryIncoming = 10000;
+  constexpr std::uint64_t kMaxTryInner = 6000;
```
We may want to declare these two as `gflags` with the default values as we set here, so that if we would like to tune the performance, we could set different values without re-compiling the code.

hakanmemisoglu replied:
Ok, I will try that one.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+
+bool LockManager::acquireLock(const transaction_id tid,
+                              const ResourceId &rid,
+                              const AccessMode access_mode) {
```
We may want to use a reference (`&`) for `access_mode` to avoid a deep copy of the object.

hakanmemisoglu replied:
It is very lightweight, just one word. Using indirection even might be slow. Can you reconsider that?

zuyu replied:
You are right that `AccessMode` is one byte.

But the reference type has the same `sizeof` value as the referenced type. So, it will not slow down anyway. And for consistency with `AccessMode::isCompatible`, I prefer to use the reference.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+    current_access_mode = (current_access_mode.isShareLock() ||
+                           current_access_mode.isIntentionShareLock())
+                              ? AccessMode(AccessModeType::kIsLock)
+                              : AccessMode(AccessModeType::kIxLock);
```
I suggest to add factory methods in `AccessMode` for all locks:

```
  static AccessMode IsLockMode() {
    return AccessMode(AccessModeType::kIsLock);
  }

  static AccessMode IxLockMode();

  ...
```

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+  if (transaction_deleted
+          == TransactionTableResult::kTransactionDeleteError) {
+    LOG(FATAL)
+        << "In LockManager.releaseAllLocks: Transaction could not be deleted!";
```
How about the following refactor:

```
  CHECK(transaction_deleted != TransactionTableResult::kTransactionDeleteError)
      << "...";
```

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+                 + std::to_string(tid)
+                 + " released lock:"
+                 + resource_id.toString()
+                 + "\n";
```
In `glog`, use `<<` instead of `+` to avoid temp string objects. Also, remove `\n`, as `glog` prints things in one line per log.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+                 + resource_id.toString()
+                 + "\n";
+
+    if (lock_deleted == LockTableResult::kDeleteError) {
```
Suggest to use `CHECK` from `glog`.

zuyu commented on transaction/LockManager.cpp:
```diff
@@ -0,0 +1,217 @@
+  CHECK(lock_result != LockTableResult::kPutError)
+      << "Unexpected result in LockManager.acquireLockInternal";
+
+  if (lock_result == LockTableResult::kAlreadyInOwned) {
```
How about rewrite this `if-else` part in a `switch`?

hakanmemisoglu replied:
switch statement gives errors in GCC. I suppose CHECK exits when it finds an unexpected condition. It confuses GCC if it is used with switch.

zuyu replied:
I was wondering if you missed the brackets for cases `LockTableResult::kPlacedInOwned` and `LockTableResult::kPlacedInPending`. The following code passes on my Mac using `clang`:

```
  switch (lock_result) {
    case LockTableResult::kAlreadyInOwned: {
      return true;
    }
    case LockTableResult::kPlacedInOwned: {
      const TransactionTableResult transaction_result
          = transaction_table_->putOwnEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInOwned)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.1).";
      return true;
    }
    case LockTableResult::kPlacedInPending: {
      const TransactionTableResult transaction_result =
        transaction_table_->putPendingEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInPending)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.2).";
      return false;
    }
    case LockTableResult::kAlreadyInPending:  // Fall through.
    default:
      return false;
  }
```

An alternative tricky implementation would be:

```
  switch (lock_result) {
    case LockTableResult::kPlacedInOwned: {
      const TransactionTableResult transaction_result
          = transaction_table_->putOwnEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInOwned)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.1).";
      // Fall through.
    }
    case LockTableResult::kAlreadyInOwned:
      return true;
    case LockTableResult::kPlacedInPending: {
      const TransactionTableResult transaction_result =
        transaction_table_->putPendingEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInPending)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.2).";
      // Fall through.
    }
    case LockTableResult::kAlreadyInPending:  // Fall through.
    default:
      return false;
  }
```

zuyu replied:
Btw, what do `No.1` and `No.2` mean in `mismatch of table results`?

hakanmemisoglu replied:
It might be the case. I will try again. Thanks.

zuyu commented on transaction/LockManager.hpp:
```diff
@@ -0,0 +1,118 @@
+#include <queue>
+#include <vector>
+
+#include "utility/ThreadSafeQueue.hpp"
```
Sort the header files in the alphabet order.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,112 @@
+class CycleDetectorTest : public testing::Test {
+ protected:
+  const std::uint64_t kNumberOfTransactions = 12;
+  virtual void SetUp() {
```
Please add one empty line before `SetUp`.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,112 @@
+      transactions_.push_back(new transaction_id(i));
+    }
+
+    wait_for_graph_ = std::make_unique<DirectedGraph>();
```
`wait_for_graph_` is initialized in each test fixture, so we could remove this here.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,112 @@
+
+  std::vector<std::pair<DirectedGraph::node_id, DirectedGraph::node_id>> edges_;
+  std::unique_ptr<DirectedGraph> wait_for_graph_;
+  std::vector<transaction_id*> transactions_;
```
Since `transactions_` is not used in other place except for `SetUp`, we could declare it as a local variable.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,112 @@
+  std::vector<std::pair<DirectedGraph::node_id, DirectedGraph::node_id>> edges_;
+  std::unique_ptr<DirectedGraph> wait_for_graph_;
+  std::vector<transaction_id*> transactions_;
+  std::vector<DirectedGraph::node_id> node_ids_;
```
Since both `transactions_` and `node_ids_` is not used in other place except for `SetUp`, we could declare them as local variables.

hakanmemisoglu replied:
But then I need to write these line in both tests.

zuyu replied:
I mean, they should become the local variables within `SetUp`, instead of a `protected` member in the test class.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,112 @@
+  EXPECT_EQ(expected_victims.size(), victims.size());
+  for (std::size_t i = 0; i < victims.size(); ++i) {
+    EXPECT_EQ(expected_victims.count(victims[i]), 1);
+  }
```
We could extract some parts of this test fixture into two methods in `CycleDetectorTest`, and then use them in all test fixtures:

```
  void initializeCycleDetector() {
    for (const auto edge : edges_) {
      wait_for_graph_->addEdgeUnchecked(edge.first, edge.second);
    }

    cycle_detector_.reset(new CycleDetector(wait_for_graph_.get()));
  }

  void checkVictims(const std::unordered_set<DirectedGraph::node_id> &expected_victims) const {
    const std::vector<DirectedGraph::node_id> victims = cycle_detector_->breakCycle();
    ASSERT_EQ(expected_victims.size(), victims.size());

    for (const auto victim : victims) {
      EXPECT_EQ(expected_victims.count(victim), 1);
    }
  }
```

zuyu commented on transaction/tests/DeadLockDetector_unittest.cpp:
```diff
@@ -0,0 +1,92 @@
+                       AccessMode(AccessModeType::kXLock));
+
+  // Run deadlock detector.
+  DeadLockDetector dl_detector(lock_table_.get(), &status_, &victims_);
```
Please use either `deadlock_detector` or `detector`. `dl` is not a good shorten expression in any case.

zuyu commented:
@hakanmemisoglu Just done another pass. Will merge after fixing these comments. Thanks!

hakanmemisoglu commented:
Thanks for the review, I applied the changes.

zuyu commented:
@hakanmemisoglu FYI, `CycleDetector_unittest` fails on my Mac for commit 41e0ae952ce55421fdf8656c140560b277d041c3.

```
[ RUN      ] CycleDetectorTest.BreakCycle
/Users/zzhang/workspace/ll/quickstep/transaction/tests/CycleDetector_unittest.cpp:65: Failure
      Expected: expected_victims.size()
      Which is: 7
To be equal to: victims.size()
      Which is: 8
[  FAILED  ] CycleDetectorTest.BreakCycle (0 ms)
```

zuyu commented:
Also, please run `IWYU` for all files under `transaction`. There are tons of issues need to fix.

One of interesting things, is that `IWYU` finds a undeclared type `TransactionId` in `Transaction.cpp`. And it seems that we do not use that file at all. So please remove it at all.

Here is the result I got from `IWYU`. Please fix them wisely. We could ignore few hints for systems headers like `<__hash_table>`, `<__bit_reference>`, and `<ostream>`. Also, if we include `threading/Thread.hpp` for example, we should ignore its suggestion to include `threading/.../Thread.hpp` instead.
```
(transaction/AccessMode.hpp has correct #includes/fwd-decls)

(transaction/AccessMode.cpp has correct #includes/fwd-decls)

transaction/CycleDetector.hpp should add these lines:
#include <__hash_table>                                 // for __hash_const_...
#include <unordered_set>                                // for unordered_set

transaction/CycleDetector.hpp should remove these lines:
- #include <set>  // lines 22-22

The full include-list for transaction/CycleDetector.hpp:
#include <__hash_table>                                 // for __hash_const_...
#include <memory>                                       // for unique_ptr
#include <unordered_set>                                // for unordered_set
#include <vector>                                       // for vector
#include "transaction/DirectedGraph.hpp"                // for DirectedGraph
#include "transaction/StronglyConnectedComponents.hpp"  // for StronglyConne...
#include "utility/Macros.hpp"                           // for DISALLOW_COPY...
---

(transaction/CycleDetector.cpp has correct #includes/fwd-decls)

transaction/DeadLockDetector.hpp should add these lines:
#include <cstdint>                        // for int64_t
#include "threading/cpp11/Thread.hpp"     // for Thread
namespace quickstep { namespace transaction { class LockTable; } }

transaction/DeadLockDetector.hpp should remove these lines:
- #include "threading/Thread.hpp"  // lines 31-31
- #include "transaction/Lock.hpp"  // lines 27-27
- #include "transaction/LockTable.hpp"  // lines 28-28
- #include "transaction/ResourceId.hpp"  // lines 29-29

The full include-list for transaction/DeadLockDetector.hpp:
#include <atomic>                         // for atomic
#include <cstdint>                        // for int64_t
#include <memory>                         // for unique_ptr
#include <unordered_map>                  // for unordered_map
#include <vector>                         // for vector
#include "threading/cpp11/Thread.hpp"     // for Thread
#include "transaction/DirectedGraph.hpp"  // for DirectedGraph::node_id, Dir...
#include "transaction/Transaction.hpp"    // for transaction_id
namespace quickstep { namespace transaction { class LockTable; } }
---

transaction/DeadLockDetector.cpp should add these lines:
#include <__hash_table>                   // for __hash_const_iterator<>::__...
#include <ostream>                        // for operator<<
#include <ratio>                          // for ratio

transaction/DeadLockDetector.cpp should remove these lines:
- #include <algorithm>  // lines 20-20

The full include-list for transaction/DeadLockDetector.cpp:
#include "transaction/DeadLockDetector.hpp"
#include <__hash_table>                   // for __hash_const_iterator<>::__...
#include <atomic>                         // for atomic
#include <chrono>                         // for seconds
#include <memory>                         // for unique_ptr, allocator, make...
#include <ostream>                        // for operator<<
#include <ratio>                          // for ratio
#include <thread>                         // for sleep_for
#include <utility>                        // for pair
#include <vector>                         // for vector, swap
#include "glog/logging.h"                 // for COMPACT_GOOGLE_LOG_FATAL, LOG
#include "transaction/CycleDetector.hpp"  // for CycleDetector
#include "transaction/DirectedGraph.hpp"  // for DirectedGraph, DirectedGrap...
#include "transaction/LockTable.hpp"      // for LockTable, LockTable::const...
#include "transaction/Transaction.hpp"    // for transaction_id
---

transaction/LockManager.hpp should add these lines:
#include <memory>                            // for unique_ptr
#include "threading/cpp11/Thread.hpp"        // for Thread
namespace quickstep { namespace transaction { class AccessMode; } }
namespace quickstep { namespace transaction { class LockRequest; } }
namespace quickstep { namespace transaction { class LockTable; } }
namespace quickstep { namespace transaction { class ResourceId; } }
namespace quickstep { namespace transaction { class TransactionTable; } }
namespace quickstep { template <typename T> class ThreadSafeQueue; }

transaction/LockManager.hpp should remove these lines:
- #include "threading/Mutex.hpp"  // lines 25-25
- #include "threading/Thread.hpp"  // lines 26-26
- #include "transaction/AccessMode.hpp"  // lines 27-27
- #include "transaction/LockRequest.hpp"  // lines 29-29
- #include "transaction/LockTable.hpp"  // lines 30-30
- #include "transaction/ResourceId.hpp"  // lines 31-31
- #include "transaction/TransactionTable.hpp"  // lines 33-33
- #include "utility/ThreadSafeQueue.hpp"  // lines 34-34

The full include-list for transaction/LockManager.hpp:
#include <atomic>                            // for atomic
#include <memory>                            // for unique_ptr
#include <queue>                             // for queue
#include <vector>                            // for vector
#include "threading/cpp11/Thread.hpp"        // for Thread
#include "transaction/DeadLockDetector.hpp"  // for DeadLockDetector (ptr only)
#include "transaction/Transaction.hpp"       // for transaction_id
namespace quickstep { namespace transaction { class AccessMode; } }
namespace quickstep { namespace transaction { class LockRequest; } }
namespace quickstep { namespace transaction { class LockTable; } }
namespace quickstep { namespace transaction { class ResourceId; } }
namespace quickstep { namespace transaction { class TransactionTable; } }
namespace quickstep { template <typename T> class ThreadSafeQueue; }
---

transaction/LockManager.cpp should add these lines:
#include <cstdint>                           // for uint64_t
#include <ostream>                           // for operator<<, basic_ostream
#include <string>                            // for char_traits, to_string

transaction/LockManager.cpp should remove these lines:
- #include "threading/Mutex.hpp"  // lines 24-24
- #include "threading/Thread.hpp"  // lines 25-25

The full include-list for transaction/LockManager.cpp:
#include "transaction/LockManager.hpp"
#include <cstdint>                           // for uint64_t
#include <ostream>                           // for operator<<, basic_ostream
#include <stack>                             // for stack
#include <string>                            // for char_traits, to_string
#include <utility>                           // for pair, make_pair
#include <vector>                            // for allocator, vector
#include "gflags/gflags.h"                   // for DEFINE_uint64
#include "glog/logging.h"                    // for CHECK, COMPACT_GOOGLE_LO...
#include "transaction/AccessMode.hpp"        // for AccessMode
#include "transaction/DeadLockDetector.hpp"  // for DeadLockDetector, DeadLo...
#include "transaction/LockRequest.hpp"       // for LockRequest, RequestType
#include "transaction/LockTable.hpp"         // for LockTable, LockTableResult
#include "transaction/ResourceId.hpp"        // for ResourceId
#include "transaction/Transaction.hpp"       // for transaction_id
#include "transaction/TransactionTable.hpp"  // for TransactionTable, Transa...
#include "utility/ThreadSafeQueue.hpp"       // for ThreadSafeQueue
---

transaction/LockTable.hpp should add these lines:
#include <memory>                           // for allocator
#include "threading/cpp11/SharedMutex.hpp"  // for SharedMutex
namespace quickstep { namespace transaction { class AccessMode; } }

transaction/LockTable.hpp should remove these lines:
- #include "threading/SharedMutex.hpp"  // lines 25-25
- #include "transaction/AccessMode.hpp"  // lines 26-26

The full include-list for transaction/LockTable.hpp:
#include <list>                             // for list
#include <memory>                           // for allocator
#include <unordered_map>                    // for unordered_map
#include <utility>                          // for pair
#include "threading/cpp11/SharedMutex.hpp"  // for SharedMutex
#include "transaction/Lock.hpp"             // for Lock
#include "transaction/ResourceId.hpp"       // for ResourceId, ResourceId::R...
#include "transaction/Transaction.hpp"      // for transaction_id
#include "utility/Macros.hpp"               // for DISALLOW_COPY_AND_ASSIGN
namespace quickstep { namespace transaction { class AccessMode; } }
---

transaction/LockTable.cpp should add these lines:

transaction/LockTable.cpp should remove these lines:
- #include <unordered_map>  // lines 21-21
- #include "threading/SharedMutex.hpp"  // lines 24-24
- #include "transaction/ResourceId.hpp"  // lines 27-27
- #include "utility/Macros.hpp"  // lines 29-29

The full include-list for transaction/LockTable.cpp:
#include "transaction/LockTable.hpp"
#include <list>                         // for __list_const_iterator, operat...
#include <utility>                      // for pair, make_pair
#include "transaction/AccessMode.hpp"   // for AccessMode
#include "transaction/Lock.hpp"         // for Lock
#include "transaction/Transaction.hpp"  // for transaction_id
---

transaction/ResourceId.hpp should add these lines:

transaction/ResourceId.hpp should remove these lines:
- #include <limits>  // lines 22-22

The full include-list for transaction/ResourceId.hpp:
#include <cstddef>                       // for size_t
#include <string>                        // for string
#include "catalog/CatalogTypedefs.hpp"   // for database_id, kInvalidCatalogId
#include "storage/StorageBlockInfo.hpp"  // for block_id, tuple_id, kInvalid...
---

transaction/ResourceId.cpp should add these lines:
#include <memory>                        // for __scalar_hash
#include <ostream>                       // for operator<<

transaction/ResourceId.cpp should remove these lines:

The full include-list for transaction/ResourceId.cpp:
#include "transaction/ResourceId.hpp"
#include <cstddef>                       // for size_t
#include <functional>                    // for hash
#include <memory>                        // for __scalar_hash
#include <ostream>                       // for operator<<
#include <string>                        // for allocator, operator+, basic_...
#include "catalog/CatalogTypedefs.hpp"   // for database_id, relation_id
#include "glog/logging.h"                // for COMPACT_GOOGLE_LOG_FATAL, LOG
#include "storage/StorageBlockInfo.hpp"  // for block_id, tuple_id
#include "utility/HashPair.hpp"          // for CombineHashes
---

(transaction/StronglyConnectedComponents.hpp has correct #includes/fwd-decls)

transaction/StronglyConnectedComponents.cpp should add these lines:
#include <__bit_reference>                // for __bit_reference

transaction/StronglyConnectedComponents.cpp should remove these lines:
- #include <cstddef>  // lines 20-20

The full include-list for transaction/StronglyConnectedComponents.cpp:
#include "transaction/StronglyConnectedComponents.hpp"
#include <__bit_reference>                // for __bit_reference
#include <cstdint>                        // for uint64_t
#include <stack>                          // for stack
#include <unordered_map>                  // for unordered_map
#include <vector>                         // for vector, allocator, vector<>...
#include "transaction/DirectedGraph.hpp"  // for DirectedGraph::node_id, Dir...
---

transaction/TransactionTable.hpp should add these lines:
namespace quickstep { namespace transaction { class AccessMode; } }

transaction/TransactionTable.hpp should remove these lines:
- #include "transaction/AccessMode.hpp"  // lines 26-26

The full include-list for transaction/TransactionTable.hpp:
#include <list>                         // for list
#include <unordered_map>                // for unordered_map
#include <utility>                      // for pair
#include <vector>                       // for allocator, vector
#include "transaction/Lock.hpp"         // for Lock
#include "transaction/ResourceId.hpp"   // for ResourceId
#include "transaction/Transaction.hpp"  // for transaction_id
#include "utility/Macros.hpp"           // for DISALLOW_COPY_AND_ASSIGN
namespace quickstep { namespace transaction { class AccessMode; } }
---

(transaction/TransactionTable.cpp has correct #includes/fwd-decls)
```

hakanmemisoglu commented:
iwyu and cpplint give conflicting results for <algorithm> and <ratio> headers. I applied cpplint suggestions. I also did not apply iwyu's false suggestions on Thread and SharedMutex classes.
I could not reproduce CycleDetector unittest failure, I will look into it.

hakanmemisoglu commented:
@zuyu, I did one pass.I still could not produce CycleDetector test fail. I suppose I am relying on an implementation detail. I will check the correctness by checking reachability of the nodes.

hakanmemisoglu commented:
@zuyu, I addressed TODO part you mentioned, you can check now. Also I changed CycleDetector test, now it will check whether all remaining nodes are not self-reachable.

zuyu commented on transaction/DeadLockDetector.hpp:
```diff
@@ -0,0 +1,156 @@
+
+#include "transaction/DirectedGraph.hpp"
+#include "transaction/Transaction.hpp"
+#include "threading/Thread.hpp"
```
Sort the header files in the alphabet order.

zuyu commented on transaction/DirectedGraph.hpp:
```diff
@@ -54,34 +54,32 @@ class DirectedGraph {
-    for (std::vector<DirectedGraphNode>::const_iterator
-           it = nodes_.cbegin(); it != nodes_.cend(); ++it) {
-      CHECK(*data != it->getData());
+  inline
```
We could remove `inline`, as every method defined in the header is implicitly `inline`d.

zuyu commented on transaction/DirectedGraph.hpp:
```diff
@@ -54,34 +54,32 @@ class DirectedGraph {
-      CHECK(*data != it->getData());
+  inline
+  node_id addNodeCheckExists(const transaction_id transaction_id_payload) {
+    for (const auto &node : nodes_) {
```
Q: I don't understand why using `vector` to represent `nodes_`? Does the insertion order matter? I would recommend `unordered_set`.

Also, I've noticed that we use `node_id` to refer a `transaction_id` in `nodes_`. I was wondering if we could use `transaction_id` directly.

hakanmemisoglu replied:
By this way, you can use node_id as index in vector. It is also general implementation technique for graph nodes as I have seen in other places as well as in DAG.hpp.
Let's keep it this way, as later on the payload can carry additional information with transaction id.

zuyu replied:
I am more concerned with the `O(n)` performance here. In query execution DAG, the number of operator nodes in a query is not significant when comparing to that of concurrent transactions.

zuyu commented on transaction/DirectedGraph.hpp:
```diff
@@ -54,34 +54,32 @@ class DirectedGraph {
-           it = nodes_.cbegin(); it != nodes_.cend(); ++it) {
-      CHECK(*data != it->getData());
+  inline
+  node_id addNodeCheckExists(const transaction_id transaction_id_payload) {
```
Optionally, we could just use `transaction` for `transaction_id_payload`.

hakanmemisoglu replied:
Let's keep it as it its.

zuyu commented on transaction/DirectedGraph.hpp:
```diff
@@ -158,8 +156,8 @@ class DirectedGraph {
    public:
-    explicit DirectedGraphNode(transaction_id *data)
-      : data_(data) {}
+    explicit DirectedGraphNode(transaction_id payload)
```
Please add `const` for `payload`, as well in many methods in this file.

Also, I suggest to just use `transaction` instead of `payload`.

zuyu commented on transaction/DirectedGraph.hpp:
```diff
@@ -180,12 +178,12 @@ class DirectedGraph {
    private:
     // Owner pointer to transaction id.
-    std::unique_ptr<transaction_id> data_;
+    transaction_id transaction_id_payload_;
```
Please remove the comment in Line 185.

Could we refactor as following:

```
  const transaction_id transaction_;
```

zuyu commented on transaction/LockManager.hpp:
```diff
@@ -0,0 +1,129 @@
+namespace transaction {
+
+class AccessMode;
+class LockRequest;
```
We have to remove this forward declaration, and include `transaction/LockRequest.hpp` back.

 Otherwise it won't compile on my Mac, as `ThreadSafeQueue` internally could not instantiate the templatized class.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  bool isSelfReachableNode(
+      DirectedGraph::node_id start_node,
```
Please mark `const` for `start_node`.

zuyu commented on transaction/tests/CycleDetector_unittest.cpp:
```diff
@@ -0,0 +1,157 @@
+  checkVictims(expected_victims);
+}
+
+TEST_F(CycleDetectorTest, BreakCycleSimple) {
```
How about we move this test fixture before `BreakCycle`?

Also, I would suggest the following renames:
 * `BreakCycleSimple` -> `Interleaving`
 * BreakCycle -> `MultipleCycles`

zuyu commented on transaction/LockTable.cpp:
```diff
@@ -47,7 +43,7 @@ LockTable::putLock(const transaction_id tid,
        it != lock_own_list.cend(); ++it) {
     if (it->first == tid && it->second.getAccessMode() == access_mode) {
-      return LockTableResult::kALREADY_IN_OWNED;
+      return LockTableResult::kAlreadyInOwned;
```
For consistency, we should also use the reference (`&`) for `access_mode` in the function signatures, including `putLock`.

zuyu commented on transaction/tests/DeadLockDetector_unittest.cpp:
```diff
@@ -0,0 +1,94 @@
+  // X lock on resource 2.
+  lock_table_->putLock(transaction_two,
+                       resource_one,
+                       AccessMode::XLockMode());
```
What do you think of the following refactor:

```
  const AccessMode x_lock(AccessModeType::kXLock);

  lock_table_->putLock(..., x_lock);
```

zuyu commented on transaction/tests/LockRequest_unittest.cpp:
```diff
@@ -31,7 +31,7 @@ class LockRequestTest : public ::testing::Test {
       : lock_request_(transaction_id(3),
                       ResourceId(5),
-                      AccessMode(AccessModeType::kSLock),
+                      AccessMode(AccessMode::SLockMode()),
```
Refactor to `AccessMode::SLockMode(),`.

zuyu commented on transaction/tests/LockRequest_unittest.cpp:
```diff
@@ -41,7 +41,7 @@ class LockRequestTest : public ::testing::Test {
   EXPECT_EQ(transaction_id(3), lock_request_.getTransactionId());
   EXPECT_EQ(ResourceId(5), lock_request_.getResourceId());
-  EXPECT_EQ(AccessMode(AccessModeType::kSLock), lock_request_.getAccessMode());
+  EXPECT_EQ(AccessMode(AccessMode::SLockMode()), lock_request_.getAccessMode());
```
Refactor to `EXPECT_EQ(AccessMode::SLockMode(), lock_request_.getAccessMode());`

zuyu commented on transaction/tests/LockTable_unittest.cpp:
```diff
@@ -43,62 +43,62 @@ class LockTableTest : public ::testing::Test {
-                                AccessMode(AccessModeType::kIsLock)),
-            LockTableResult::kPLACED_IN_PENDING);
+                                AccessMode::IsLockMode()),
+            LockTableResult::kPlacedInPending);
```
FYI, `EXPECT_EQ` has a convention that the left hand side is the expected value, and the right hand side is the checked value. So, in this case, we should refactor to `EXPECT_EQ(LockTableResult::kPlacedInPending, ...);`

Please do such reverses in this file.

Also, I suggest to declare `const AccessMode is_lock(AccessModeType::kIsLock), x_lock(AccessModeType::kXLock);`, and then use the variables in `putLock`.

zuyu commented on transaction/tests/StronglyConnectedComponents_unittest.cpp:
```diff
@@ -43,7 +43,7 @@ class GraphConfiguration {
       std::unique_ptr<transaction_id> tid =
         std::make_unique<transaction_id>(transaction_id(index));
```
We should refactor to

```
  const transaction_id transaction = static_cast<transaction_id>(index);
```

Also, rename `no_transaction` to `num_transactions`.

zuyu commented on transaction/tests/TransactionTable_unittest.cpp:
```diff
@@ -116,15 +116,15 @@ TEST_F(TransactionTableTest, TransactionEntries) {
   EXPECT_EQ(transaction_table_.deleteOwnEntry(tid_1_,
                                               ResourceId(4),
-                                              AccessMode(AccessModeType::kSLock)),
+                                              AccessMode::SLockMode()),
```
Since we are only using two locks, we could define both `s_lock` and `x_lock`.

zuyu commented:
@hakanmemisoglu Good job! All unit tests pass on my Mac, given that we include `LockRequest` header file.

There are some minor issues to fix before merging.

hakanmemisoglu commented:
For your latest linear time concern:
I did not used this method anywhere in the module. I  added this method after @hbdeshmukh recommended me. The versions with check are there for solely debug purpose, if someone else wants to use later on. Linear time for that kind of question is normal: Is particular payload is in the graph? You cannot answer this question efficiently, unless you keep additional information in the graph, or use another underlying data structure. In this module this information data->node_index mapping is tracked by DeadLockDetector.

zuyu commented:
I would say that linear time is not normal for this case, the constant time is, i.e., using `unordered_set`.

Since we may add more info in the `node` and that method is only for debug, it is ok to leave it as is for now. But we could have a comment stating that we will use efficient data structures later.

zuyu commented on transaction/tests/StronglyConnectedComponents_unittest.cpp:
```diff
@@ -35,11 +35,11 @@ namespace transaction {
  public:
   GraphConfiguration(DirectedGraph *graph,
-                     std::size_t no_transaction,
+                     std::size_t num_transactions,
```
Please mark `const`.
